### PR TITLE
feat: new module avm/res/managed-identity/user-assigned-identity

### PR DESCRIPTION
**Pull request by elanzel**
## Description

New module migrated from CARML to AVM

[avm.res.managed-identity.user-assigned-identity](https://github.com/elanzel/bicep-registry-modules/actions/workflows/avm.res.managed-identity.user-assigned-identity.yml)

### CODE REVIEW

File: avm/res/managed-identity/user-assigned-identity/main.bicep [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  }
+}
+
+resource userAssignedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' = {
+  name: name
+  location: location
+  tags: tags
+}
+
+resource userMsi_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
```
Comment (eriqua) [inline avm/res/managed-identity/user-assigned-identity/main.bicep:60]:
```suggestion
resource userAssignedIdentity_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock ?? {}) && lock.?kind != 'None') {
```

File: avm/res/managed-identity/user-assigned-identity/main.bicep [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  name: lock.?name ?? 'lock-${name}'
+  properties: {
+    level: lock.?kind ?? ''
+    notes: lock.?kind == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot delete or modify the resource or child resources.'
+  }
+  scope: userAssignedIdentity
+}
+
+module userMsi_federatedIdentityCredentials 'federated-identity-credential/main.bicep' = [for (federatedIdentityCredential, index) in federatedIdentityCredentials: {
```
Comment (eriqua) [inline avm/res/managed-identity/user-assigned-identity/main.bicep:69]:
```suggestion
module userAssignedIdentity_federatedIdentityCredentials 'federated-identity-credential/main.bicep' = [for (federatedIdentityCredential, index) in federatedIdentityCredentials: {
```

File: avm/res/managed-identity/user-assigned-identity/main.bicep [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  params: {
+    name: federatedIdentityCredential.name
+    userAssignedIdentityName: userAssignedIdentity.name
+    audiences: federatedIdentityCredential.audiences
+    issuer: federatedIdentityCredential.issuer
+    subject: federatedIdentityCredential.subject
+  }
+}]
+
+resource userMsi_roleAssignments 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for (roleAssignment, index) in (roleAssignments ?? []): {
```
Comment (eriqua) [inline avm/res/managed-identity/user-assigned-identity/main.bicep:80]:
```suggestion
resource userAssignedIdentity_roleAssignments 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for (roleAssignment, index) in (roleAssignments ?? []): {
```

**Review by skydiver1980 — APPROVED** [review]

**Review by skydiver1980 — APPROVED** [review]

File: avm/res/managed-identity/user-assigned-identity/main.bicep [hunk 4]
```diff
@@ -0,0 +1,145 @@
+metadata name = 'User Assigned Identities'
+metadata description = 'This module deploys a User Assigned Identity.'
+metadata owner = 'Azure/module-maintainers'
+
+@description('Optional. Name of the User Assigned Identity.')
+param name string = guid(resourceGroup().id)
```
Comment (AlexanderSehr) [inline avm/res/managed-identity/user-assigned-identity/main.bicep:6]:
```suggestion
@description('Required. Name of the User Assigned Identity.')
param name string
```

@eriqua would you agree we should not have this default value?

Comment (AlexanderSehr) [reply avm/res/managed-identity/user-assigned-identity/main.bicep:6]: Note: This was already implemented like this (i.e., with the default value in CARML)

File: avm/res/managed-identity/user-assigned-identity/tests/e2e/defaults/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+// ============== //
+
+@batchSize(1)
+module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
+  params: {
+    location:location
+    // Workaround for PSRule
+    tags: null
```
Comment (AlexanderSehr) [inline avm/res/managed-identity/user-assigned-identity/tests/e2e/defaults/main.test.bicep:45]:
```suggestion
    location: location
```
Not required anymore

File: avm/res/managed-identity/user-assigned-identity/tests/e2e/waf-aligned/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+module nestedDependencies 'dependencies.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    location: location
+    managedIdentityName: 'dep-${namePrefix}-msi-${serviceShort}'
+  }
+}
+
```
Comment (AlexanderSehr) [inline avm/res/managed-identity/user-assigned-identity/tests/e2e/waf-aligned/main.test.bicep:42]:
```suggestion
```

Unless I'm mistaken, this is not used & the file can be removed

File: avm/res/managed-identity/user-assigned-identity/main.bicep (hunk 4 above)
Comment (eriqua) [reply avm/res/managed-identity/user-assigned-identity/main.bicep:6]: Agree, sorry I didn't catch this and the other 2 points below. Thanks!

**Review by eriqua — DISMISSED** [review]
🆗

File: .github/workflows/avm.res.managed-identity.user-assigned-identity.yml [hunk 7]
```diff
@@ -0,0 +1,81 @@
+name: "avm.res.managed-identity.user-assigned-identity"
```
Comment (eriqua) [inline .github/workflows/avm.res.managed-identity.user-assigned-identity.yml:1]: @elanzel as discussed offline please also make sure of the following:
- update codeowners file to uncomment the line referencing your module
- make sure your module owner team is properly added to the intended parent team CC: @matebarabas

**Review by eriqua — APPROVED** [review]
:ok:

**Pull request merged**